### PR TITLE
MSA: Cleanup SimulationWidget

### PR DESCRIPTION
**Pull request by rhaschke**
`config_data_->save_gazebo_urdf_` is True if and only if `config_data_->changes` has the `MoveItConfigData::SIMULATION` bit set. Thus we can replace the former with the latter check.
Fixes #3280

### CODE REVIEW

**Review by BobbyCephy — COMMENTED** [review]
outputGazeboURDFFile might remove an empty URDF file with this patch:
[Delete empty Gazebo URDF file.txt](https://github.com/ros-planning/moveit/files/10086105/Delete.empty.Gazebo.URDF.file.txt)

The GAZEBO_URDF_LOAD_ATTRIBUTE still has to be updated, if the urdf is unchecked in the configuration files widget with this patch:
[Update GAZEBO_URDF_LOAD_ATTRIBUTE by checkbox.txt](https://github.com/ros-planning/moveit/files/10086140/Update.GAZEBO_URDF_LOAD_ATTRIBUTE.by.checkbox.txt)

The initial state of the urdf checkbox should be set with this patch:
[Set initial checkbox state.txt](https://github.com/ros-planning/moveit/files/10086357/Set.initial.checkbox.state.txt)

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 1]
```diff
@@ -964,8 +964,18 @@ bool ConfigurationFilesWidget::checkGenFiles()
           ROS_WARN_STREAM("Editing in Setup Assistant conflicts with external editing of file " << file->file_name_);
 
         file->generate_ = false;  // do not overwrite by default
-        file->modified_ = true;
-        found_modified = true;
+
```
Comment (rhaschke) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:969]: Why do you need this change? As far as I understand, this enforces overwriting the file even if it has changes since last MSA generation.

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 2]
```diff
@@ -285,6 +285,7 @@ bool ConfigurationFilesWidget::loadGenFiles()
       "for more infos.";
   file.gen_func_ = [this](const std::string& output_path) { return config_data_->outputGazeboURDFFile(output_path); };
   file.write_on_changes = MoveItConfigData::SIMULATION;
+  file.generate_ = config_data_->changes & MoveItConfigData::SIMULATION;
```
Comment (rhaschke) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:288]: Why is that needed?

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 3]
```diff
@@ -300,6 +301,7 @@ bool ConfigurationFilesWidget::loadGenFiles()
                       "parameters can be defined but are optional.";
   file.gen_func_ = [this](const std::string& output_path) { return config_data_->outputOMPLPlanningYAML(output_path); };
   file.write_on_changes = MoveItConfigData::GROUPS;
+  file.generate_ = true;
```
Comment (rhaschke) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:304]: Same here.

**Review by rhaschke — COMMENTED** [review]
Dear @BobbyCephy

next time, please file a new PR with your additional suggestions and even more important: explain them. That's much easier for you as well as for us maintainers. I applied the changes from your text file.

I don't yet understand your last two commits (see below).
Please also [provide an official review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews). Thanks.

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 4]
```diff
@@ -967,8 +966,18 @@ bool ConfigurationFilesWidget::checkGenFiles()
...
+        if (file->file_name_.find("urdf") == std::string::npos)
+        {
+          file->modified_ = true;
+          found_modified = true;
+        }
+
+        else
+        {
+          file->modified_ = false;
+        }
```
Comment (BobbyCephy) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:980]: Revert
```suggestion
        file->modified_ = true;
        found_modified = true;
```

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 5]
```diff
@@ -1130,6 +1139,8 @@ bool ConfigurationFilesWidget::generatePackage()
     }
   }
 
+  loadTemplateStrings();
+
```
[the remark this answers is not in the document]
Comment (BobbyCephy) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:1143]: They might be cleared here:
https://github.com/ros-planning/moveit/blob/431fa5b159a860ada99192b800ab75df6c6b306b/moveit_setup_assistant/src/widgets/configuration_files_widget.cpp#L1252-L1253
```
  // Clear template strings in case this function is run multiple times with changes in between
  template_strings_.clear();
```

**Review by BobbyCephy — CHANGES_REQUESTED** [review]
If you know, that the urdf is only generated having opened the _Simulation_ tab, you now can choose between saving it in the config folder or generating it in launch files. But this decision should be made clearer by:
- [ ] Always generating the urdf content without opening the _Simulation_ tab and (un)checking the urdf file in the _Configuration Files_ tab, to (delete or) save it.
- [ ] Adding a checkbox in the _Simulation_ tab for saving or generating the urdf in launch files and apply this selection by checking the urdf file in the _Configuration Files_ tab.

File: moveit_setup_assistant/src/tools/moveit_config_data.cpp [hunk 6]
```diff
@@ -210,18 +210,24 @@ bool MoveItConfigData::outputSetupAssistantFile(const std::string& file_path)
...
+// Output Gazebo URDF file (or delete existing file if gazebo_urdf_string_ is empty)
 // ******************************************************************************************
 bool MoveItConfigData::outputGazeboURDFFile(const std::string& file_path)
 {
+  if (gazebo_urdf_string_.empty())
+  {
+    fs::remove(file_path.c_str());
+    return true;
+  }
+
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_setup_assistant/src/tools/moveit_config_data.cpp:222]: This is taken into account by the early return in the `fs::remove` branch.

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 7]
```diff
@@ -276,19 +276,17 @@ bool ConfigurationFilesWidget::loadGenFiles()
   config_data_->srdf_pkg_relative_path_ = file.rel_path_;
 
   // gazebo_<ROBOT>.urdf ---------------------------------------------------------------------------------------
-  if (config_data_->save_gazebo_urdf_)
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:279]: Good point. Fixed now.

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h [hunk 8]
```diff
@@ -235,9 +235,6 @@ class MoveItConfigData
   // NOTE: Created when the robot urdf is not compatible with Gazebo.
   std::string gazebo_urdf_string_;
 
-  /// Whether a new Gazebo URDF is created
-  bool save_gazebo_urdf_;
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:239]: I don't understand this comment. The gazebo_<ROBOT>.urdf is _only_ generated in the `config` folder.
There is a switch though for the `[GAZEBO_URDF_LOAD_ATTRIBUTE]` template string to point to the `gazebo_<ROBOT>.urdf` or the default `<ROBOT>.urdf`.

[the remark this answers is not in the document]
Comment (BobbyCephy) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:239]: I got it, but should this switch not be more clearly than opening the simulation tab or not?

**Comment by rhaschke** [comment]
@BobbyCephy, I found some time to look deeper into all this and pushed a bunch of new commits.
From my point of view the following behavior should be implemented now:
- If the URDF doesn't need any changes for Gazebo, this is indicated and `gazebo_*.urdf` is not written.
  The `demo_gazebo.launch` uses the default URDF.
- Otherwise, the user can choose from these options:
  - Overwrite the existing URDF with the suggested changes (but that's only possible if `xacro` wasn't used).
    On success, the "overwrite" button is disabled and, again, no `gazebo_*.urdf` is written.
    On failure, `gazebo_*.urdf` would be written as a fallback.
  - Open the existing URDF and manually edit it. Currently, there is no re-checking of the (possibly changed) URDF implemented.
- In any other case, i.e. having changes to the URDF, a new file `gazebo_*.urdf` is written and used in `demo_gazebo.launch`

File: moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:239]: I think the reasoning behind that was: only if a user visited that tab, she seems to be interested in simulation...
You can still decide on the "package generation" tab, which files to actually generate.
The switch was just needed to distinguish between default .urdf and MSA-generated gazebo_*.urdf.
See https://github.com/ros-planning/moveit/pull/3281#issuecomment-1328040853 for a summary of the intended behavior.

**Comment by rhaschke** [comment]
@BobbyCephy, if you verified the described behavior and confirmed that your issues are all resolved, please approve.

**Review by BobbyCephy — APPROVED** [review]

**Review by BobbyCephy — APPROVED** [review]
